python/benchmark/test_gen_data.py

```diff
@@ -257,13 +365,70 @@ def test_make_sparse_regression(
         total = row_num * col_num
 
         # If there is no random shuffled redundant cols, we can check the total density
-        density_num = float(density)
-        if redundant_cols == "0":
+        if redundant_cols == "0" and density_curve == "None":
+            if isinstance(density, List):
+                density_num = sum([float(d) for d in density]) / len(density)
+            else:
+                density_num = float(density)
+
             assert (
                 count > total * density_num * 0.95
                 and count < total * density_num * 1.05
             )
 
+        # If no shuffle, test to see if the chunk density is as specified/curved
+        n_chunks_num = int(n_chunks)
+        if shuffle == "False":
+            orig_cols = col_num - int(redundant_cols)
+            num_partitions = 3
+
+            if isinstance(density, List):
+                density_num = float(density[0])
+            else:
+                density_num = float(density)
+
+            if density_curve == "Linear":
+                density_values = np.linspace(
+                    num_partitions / row_num, density_num, n_chunks_num
```

Comment:
why density_values start from "num_partitions / row_num" not 0.? Can you give some explanation?

Reply:
The starting point for this is that I want to gaurantee the smallest density for each partition is big enough to generate at least one dense entry for the column. Thus, the density need to be at least 1/(rows/num_par) = num_par/rows. One potential issue is that num_partition is larger than row_num but that would defeat the whole purpose of this distributed datagen, so it should not happen at all. I wonder what's the "not 0." issue?